src/zkscript/transaction_introspection/transaction_introspection.py

```diff
@@ -0,0 +1,220 @@
+"""Bitcoin scripts that achieve transaction introspection."""
+
+from tx_engine import SIGHASH, Script, encode_num, hash256d
+from tx_engine.engine.util import GROUP_ORDER_INT, Gx, Gx_bytes
+
+from src.zkscript.types.stack_elements import StackBaseElement
+from src.zkscript.util.utility_scripts import (
+    bool_to_moving_function,
+    bytes_to_unsigned,
+    int_sig_to_s_component,
+    move,
+    nums_to_script,
+    pick,
+    reverse_endianness,
+    roll,
+)
+
+pushtx_bit_shift_data = {
+    2: {
+        "signature_prefix": bytes.fromhex("3045022100"),
+        # R = 2^2 * G
+        "R": bytes.fromhex("02e493dbf1c10d80f3581e4904930b1404cc6c13900ee0758474fa94abe8c4cd13"),
+        # P = a * G s.t. a * R_x = -1 mod GROUP_ORDER_INT
+        "P": bytes.fromhex("034218426b38c75b706db9010aad7795fd05b872060921c048d9a679d8878c7660"),
+    },
+    3: {
+        "signature_prefix": bytes.fromhex("30440220"),
+        # R = 2^3 * G
+        "R": bytes.fromhex("022f01e5e15cca351daff3843fb70f3c2f0a1bdd05e5af888a67784ef3e10a2a01"),
+        # P = a * G s.t. a * R_x = -1 mod GROUP_ORDER_INT
+        "P": bytes.fromhex("03ad36fad55727ebf76f8af96c7c2df9a298dc21d6c15269fdedfd47a70b327637"),
+    },
+}
+
+
+class TransactionIntrospection:
+    """Class generating Bitcoin scripts that achieve transaction introspection."""
+
+    @staticmethod
+    def pushtx(
+        sighash_value: SIGHASH,
+        sig_hash_preimage: StackBaseElement = StackBaseElement(0),  # noqa: B008
+        rolling_option: bool = True,
+        clean_constants: bool = True,
+        verify_constants: bool = True,
+        is_checksigverify: bool = True,
+        is_opcodeseparator: bool = False,
+    ) -> Script:
+        """Construct PUSHTX locking script.
+
+        Stack input:
+            - stack:    [GROUP_ORDER_INT, Gx, Gx_bytes, .., sig_hash_preimage, ..]
+            - altstack: []
+        Stack output:
+            - stack:    [GROUP_ORDER_INT, Gx, Gx_bytes, .., sig_hash_preimage, ..] or fail
```

Comment:
Since there are so many optional arguments, I would probably try describe the output using the default arguments (for instance, cleaning the constant in the alt stack).

Reply:
Discussed and addressed